File: src/main/java/com/raychenon/leetcode/FloodFill.java
package com.raychenon.leetcode;

import java.util.LinkedList;
import java.util.Queue;

/**
 * User: raychenon
 * Date: 2019-09-18
 * https://github.com/raychenon/algorithms/issues/97
 */
public class FloodFill {

    public int[][] floodFill(int[][] image, int row, int col, int newColor) {
        int currentColor = image[row][col];
        if (currentColor != newColor) DFS(image, row, col, currentColor, newColor);
        return image;
    }

    private void DFS(int[][] image, int row, int col, int color, int targetColor) {
        if (image[row][col] == color) {
            image[row][col] = targetColor;
            if (row > 0) DFS(image, row - 1, col, color, targetColor); // top
            if (col > 0) DFS(image, row, col - 1, color, targetColor); // left
            if (row < image.length - 1) DFS(image, row + 1, col, color, targetColor);     // bottom
            if (col < image[0].length - 1) DFS(image, row, col + 1, color, targetColor);  // right
        }
    }

    public int[][] floodFillIterative(int[][] image, int sr, int sc, int newColor) {
        int oldColor = image[sr][sc];
        if (oldColor == newColor) return image;
        int[] pairs = {1, 0, -1, 0, 1};
        Queue<int[]> q = new LinkedList<>();
        image[sr][sc] = newColor;
        q.add(new int[]{sr, sc});
        while (!q.isEmpty()) {
            int[] rc = q.poll();
            for (int i = 1; i < pairs.length; i++) {
                int r = rc[0] + pairs[i - 1], c = rc[1] + pairs[i];
                if (r < 0 || r >= image.length || c < 0 || c >= image[0].length || image[r][c] != oldColor) continue;
                image[r][c] = newColor;
                q.add(new int[]{r, c});
            }
        }
        return image;
    }

}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/FloodFill.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 2019-09-18
 * https://leetcode.com/problems/flood-fill
 */
object FloodFill {

    fun floodFill(image: Array<IntArray>, row: Int, col: Int, targetColor: Int): Array<IntArray> {
        val color = image[row][col]
        if (color != targetColor) DFS(image, row, col, color, targetColor)
        return image
    }

    private fun DFS(image: Array<IntArray>, row: Int, col: Int, fromColor: Int, targetColor: Int): Unit {
        if (image[row][col] == fromColor) {
            image[row][col] = targetColor
            if (row > 0) DFS(image, row - 1, col, fromColor, targetColor)                  // top
            if (col > 0) DFS(image, row, col - 1, fromColor, targetColor)                   // left
            if (row < image.size - 1) DFS(image, row + 1, col, fromColor, targetColor)     // bottom
            if (col < image[0].size - 1) DFS(image, row, col + 1, fromColor, targetColor)   // right
        }
    }


    fun floodFillRec(image: Array<IntArray>, sr: Int, sc: Int, newColor: Int): Array<IntArray> {
        if(image[sr][sc] == newColor) return image
        return floodFill(image,sr,sc,image[sr][sc], newColor)
    }

    private fun floodFill(image: Array<IntArray>, sr: Int, sc: Int,oldColor: Int, newColor: Int): Array<IntArray> {
        // outside of boundaries and cell is different color
        if(sr < 0 || sr >= image.size || sc < 0 || sc >= image[0].size ||
            image[sr][sc] != oldColor) return image

        image[sr][sc] = newColor

        floodFill(image, sr+1,sc, oldColor, newColor)
        floodFill(image, sr-1,sc, oldColor, newColor)
        floodFill(image, sr,sc+1, oldColor, newColor)
        floodFill(image, sr,sc-1, oldColor, newColor)

        return image
    }
    
}